datasource/ethereum/src/block_ingestor.rs

```diff
@@ -304,61 +191,23 @@ where
     fn get_blocks<'a>(
         &'a self,
         block_hashes: &[H256],
-    ) -> Box<Stream<Item = EthereumBlock, Error = BlockIngestorError> + Send + 'a> {
-        // Don't bother with a batch request if nothing to request
-        if block_hashes.is_empty() {
-            return Box::new(stream::empty());
-        }
-
-        let batching_web3 = Web3::new(Batch::new(self.web3_transport.clone()));
-
-        // Add requests to batch
-        let block_futures = block_hashes
-            .into_iter()
-            .map(|block_hash| {
-                let block_hash = block_hash.to_owned();
-
-                batching_web3
-                    .eth()
-                    .block_with_txs(BlockId::from(block_hash))
-                    .map_err(|e| format_err!("could not get block from Ethereum: {}", e).into())
-                    .and_then(move |block_opt| {
-                        block_opt.ok_or_else(|| BlockIngestorError::BlockUnavailable(block_hash))
-                    })
-                    .and_then(move |block| self.load_full_block(block))
-            })
-            // Collect to ensure that `block_with_txs` calls happen before `submit_batch`
-            .collect::<Vec<_>>();
-
-        // Submit all requests in batch
-        Box::new(
-            batching_web3
-                .transport()
-                .submit_batch()
-                .map_err(|e| format_err!("could not get blocks from Ethereum: {}", e).into())
-                .map(|_| stream::futures_unordered(block_futures))
-                .flatten_stream(),
-        )
-    }
-}
+    ) -> Box<Stream<Item = EthereumBlock, Error = EthereumAdapterError> + Send + 'a> {
+        let logger = self.logger.clone();
+        let eth_adapter = self.eth_adapter.clone();
 
-#[derive(Debug, Fail)]
-enum BlockIngestorError {
-    /// The Ethereum node does not know about this block for some reason, probably because it
-    /// disappeared in a chain reorg.
-    #[fail(
-        display = "block data unavailable, block was likely uncled (block hash = {:?})",
-        _0
-    )]
-    BlockUnavailable(H256),
+        let block_futures = block_hashes.iter().map(move |&block_hash| {
+            let logger = logger.clone();
+            let eth_adapter = eth_adapter.clone();
 
-    /// An unexpected error occurred.
-    #[fail(display = "error in block ingestor: {}", _0)]
-    Unknown(Error),
-}
+            eth_adapter
+                .block_by_hash(&logger, block_hash)
+                .from_err()
+                .and_then(move |block_opt| {
+                    block_opt.ok_or_else(|| EthereumAdapterError::BlockUnavailable(block_hash))
+                })
+                .and_then(move |block| eth_adapter.load_full_block(&logger, block))
```

Comment:
It would be nice to have a generic `full_blocks_by_hashes` method on EthereumAdapter though that does batching of the eth_getBlockByHash and eth_getTransactionReceipt calls. That would need just two batches for all the blocks. This PR just creates `load_full_block`, which does batching of the tx receipt calls for one block (there are a lot of those so that helps quite a bit). There's a bit of complexity there in handling the BlockUnavailable case appropriately, and there might be some issues with putting too many calls into one batch -- I haven't tested that though!

Reply:
I've ran this locally and I didn't notice any slowness. Let's get this in and see how we fare!